src/libraries/System.IO.Compression.ZipFile/tests/ZipFile.Create.Stream.cs

```diff
@@ -0,0 +1,172 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Compression.Tests;
+
+public class ZipFile_Create_Stream : ZipFileTestBase
+{
+    [Fact]
+    public void CreateFromDirectory_NullSourceDirectory_Throws()
+    {
+        using MemoryStream ms = new MemoryStream();
```

Comment:
FWIW, you don't actually need to dispose of MemoryStream.

Reply:
```suggestion
        MemoryStream ms = new MemoryStream();
```

Reply:
I see that calling the empty dispose on MemoryStream is not vaporized in codegen. Should it be/can it be? It's arguably cleaner to dispose everything disposable, eg., presumably that is helpful for some static analyzer.
cc @AndyAyersMS

Reply:
In general, the JIT should inline empty methods, if they're called directly. However it may not inline the version that is left in a residual fault clause if the `using` entails a `try {...}  finally {...}`  (some do, some don't) and the JIT decides to clone the finally.

If you can point us at a repro case either @egorbo or I will investigate.

Reply:
I was looking at https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BYAKAAEAmARjPIAYACc6gOgEkB5AbjLoGZmlRgA0yAbzKNpzakmYpGAMWoAKAJRSZk0jL3MmAN2xRG+XIwC8jAHYwA7owCyMfNACeAZQyxs+Dby6+tIAwhA2uBAANjCsAOpQAJYYMKrmrAAyMDYA5hgAFuqBegC+WtLlsvLkikqUGpU6wdLGpuZWtg7Orh7evv5FlXphEdGxCcmp6Vm5BYNB0mWkS3qVQA

(not sure how to tell what version sharplab is using)

Reply:
Here is how MemoryStream is organized:
```cs
    public abstract class Stream
    {
        public void Dispose() => Close();

        public virtual void Close()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing) {}
    }

    public class MemoryStream : Stream
    {
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                _isOpen = false;
                _writable = false;
                _expandable = false;
                _lastReadTask = default;
            }
        }
    }
```
So essentially jit can't throw Dispose away - it has to emit `GC.SuppressFinalize` and null those fields as requested.
Also, note that `MemoryStream` is not sealed so users can override it. 
We might be able to help JIT here by checking that the current object is MemoryStream (not overriden) if needed

Reply:
So perhaps we should dispose all MemoryStreams to avoid pressure on the finalization queue.

Reply:
Trying to make sure I understand: MemoryStream Dispose calls SuppressFinalize() as a perf optimization, but it's small enough that it's OK to not dispose?

Reply:
SuppressFinalize is a nop / pure overhead if an object doesn't have a finalizer.  MemoryStream.Dispose calls SuppressFinalize in case a type derived from MemoryStream adds a finalizer; it's following the Dispose pattern.  If you just construct a MemoryStream and not something derived from it, its Dispose does nothing useful.

Reply:
Would it be possible to add a check for the concrete type within the Dispose, so that if the type was not derived it would elide the content of Dispose, improving the code for anyone unnecessarily calling it?

Reply:
What are you trying to avoid, the virtual call to Dispose(bool)?  That's necessary in the general case as it contains all of the `x = false` lines that result in exceptions being thrown from subsequent usage, that transition CanXx properties from returning true to false, etc.

I'm not seeing the value of trying to microoptimize these calls.  It's simply the case that code doesn't need to contort itself to dispose of a MemoryStream it itself created.  It can if it likes, but it needn't.